### PR TITLE
Feature/287 user graduation detection

### PR DESCRIPTION
**Pull request by niraymak**
## Description

This branch has:
- added a new controller called UserTaskController which is responsible for user tasks.
In our case a user task is converting a user account to alumni role which needs user input to complete.

One of the endpoint is ConvertToAlumni. This endpoint checks if the user meets the requirements. When it does it makes a call to the identityserver to change login and password. If this is succesfull the account is converted to Alumni.

- added a new project called JobScheduler which is responsible for handling scheduled jobs.
The project contains a job which does a Rest request to the API. It asks for all graduating users that do not yet have a task for it.
During the call the API creates the tasks. When the notification system is ready it will be used to queue emails to expected graduating users.

## Notes

ProjectWrite scope was functionality is now changed: ProjectWrite scope is required to create projects.
Before the functionality was to edit projects that are not yours. A new scope was created for that called AdminProjectWrite.
In the seed this is fixed, but in staging and production we should manually add the scopes to the correct roles:
Add ProjectWrite scope to all roles except Alumni.
Add AdminProjectWrite scope to role Administrator.

## Type of change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] This change requires a documentation update

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I did not update API Controllers, if I did, I added/changed Postman tests
-   [x] I have performed a self-review of my own code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [ ] I have made corresponding changes to the documentation
-   [x] I have added tests that prove my fix is effective or that my feature works
-   [ ] New and existing unit tests pass locally with my changes
-   [x] I updated the changelog with an end-user readable description
-   [x] I assigned this pull request to the correct project board to update the sprint board

## Steps to Test or Reproduce

1. Run code
2. In the user table make sure there is at least one user that has ExpectedGraduationDate within now and next 6 months.
3. Make a get call to /api/UserTask
4. Receive user mentioned in step 2. 
5. Make a put call to /api/UserTask with the new credentials in the header

## Link to issue

Closes: #287

### CODE REVIEW

File: API/Controllers/UserTaskController.cs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+            this.userTaskService = userTaskService;
+            this.userService = userService;
+            this.roleService = roleService;
+        }
+
+        /// <summary>
+        /// Creates and returns all graduation user tasks for expecting graduation users.
+        /// </summary>
+        /// <returns> All user tasks which are created or open for graduation users. </returns>
+        [HttpGet]
```
Comment (Brend-Smits) [inline API/Controllers/UserTaskController.cs:67]: Missing "ProducesResponseType" documentation. Check other controllers for examples.

File: API/Controllers/UserTaskController.cs [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+                                             Title = "Alumni role does not exist.",
+                                             Detail = "The database does not contain a role for alumni.",
+                                             Instance = "2DE31766-5D9C-4E41-908F-389A9A85F723"
+                                         };
+                return NotFound(problem);
+            }
+
+            
+            // Rest call to Identity server to change credentials. Credentials are in the headers due to security issues.
+            RestClient restClient = new RestClient("https://localhost:5005/");
```
Comment (Brend-Smits) [inline API/Controllers/UserTaskController.cs:126]: Hardcoded URL

File: Data/4_Data.csproj [hunk 3]
```diff
@@ -10,6 +10,27 @@
     <EmbeddedResource Remove="Configurations\IdentityMigrations\**" />
     <None Remove="Configurations\IdentityMigrations\**" />
   </ItemGroup>
+
+  <ItemGroup>
+    <Compile Remove="Migrations\20200527095846_addIsPublicFlagToUser.cs" />
```
Comment (Brend-Smits) [inline Data/4_Data.csproj:15]: I think these can be removed (this is only relevant if you remove a migration that is actually in develop already)

File: IdentityServer/Quickstart/Account/AccountController.cs [hunk 4]
```diff
@@ -445,6 +447,32 @@ private async Task<LoggedOutViewModel> BuildLoggedOutViewModelAsync(string logou
             return vm;
         }
 
-    }
+        //[HttpPut]
+        //[ValidateAntiForgeryToken]
```
Comment (Brend-Smits) [inline IdentityServer/Quickstart/Account/AccountController.cs:451]: Commented out code

File: IdentityServer/Views/Shared/_ValidationScriptsPartial.cshtml [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    <script src="~/lib/jquery-validation-unobtrusive/jquery.validate.unobtrusive.js"></script>
+</environment>
+<environment names="Staging,Production">
+    <script src="https://ajax.aspnetcdn.com/ajax/jquery.validate/1.17.0/jquery.validate.min.js"
+            asp-fallback-src="~/lib/jquery-validation/dist/jquery.validate.min.js"
+            asp-fallback-test="window.jQuery && window.jQuery.validator"
+            crossorigin="anonymous"
+            integrity="sha384-rZfj/ogBloos6wzLGpPkkOr/gpkBNLZ6b6yLy4o+ok+t/SAKlL5mvXLr0OXNi1Hp">
+    </script>
+    <script src="https://ajax.aspnetcdn.com/ajax/jquery.validation.unobtrusive/3.2.9/jquery.validate.unobtrusive.min.js"
```
Comment (Brend-Smits) [inline IdentityServer/Views/Shared/_ValidationScriptsPartial.cshtml:12]: What is this file for? What is it's purpose?

File: IdentityServer/appsettings.json [hunk 6]
```diff
@@ -38,7 +38,11 @@
         "swagger": {
             "RedirectUrisSwagger": "",
             "PostLogoutUrisSwagger": ""
-        }
+        },
+        "JobScheduler": {
```
Comment (Brend-Smits) [inline IdentityServer/appsettings.json:42]: You will also need to add these to the appsettings.development json

File: JobScheduler/ApiRequestHandler.cs [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        private List<UserTask> userTasks;
+
+        public ApiRequestHandler(Uri baseUrlApi)
+        {
+            apiClient = new RestClient(baseUrlApi);
+        }
+
+        public string GetToken()
+        {
+            RestClient restClient = new RestClient("https://localhost:5005/");
```
Comment (Brend-Smits) [inline JobScheduler/ApiRequestHandler.cs:30]: hardcoded url

File: JobScheduler/ApiRequestHandler.cs [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+
+        public List<UserTask> GetExpectedGraduationUsers() {
+        
+            RestRequest restRequest = new RestRequest("api/UserTask") { Method = Method.GET };
+
+            if(accessToken == null)
+            {
+                
+                dynamic data = JObject.Parse(GetToken());
```
Comment (Brend-Smits) [inline JobScheduler/ApiRequestHandler.cs:55]: It's better create a token model

File: JobScheduler/Config.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+using Services.Sources;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace JobScheduler
+{
+    public class Config
+    {
+        const string adress = "https://localhost:5005/";
```
Comment (Brend-Smits) [inline JobScheduler/Config.cs:12]: Hardcoded URL

File: JobScheduler/Config.cs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+        public string GetJwtToken()
```
Comment (Brend-Smits) [inline JobScheduler/Config.cs:14]: What happens when a token has expired? Remember, they have expiration dates!

File: JobScheduler/Config.cs [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+            RestRequest restRequest = new RestRequest("connect/token", Method.POST);
+            restRequest.AddParameter("application/json", credentialsTokenRequest, ParameterType.RequestBody);
+            restRequest.RequestFormat = DataFormat.Json;
+
+            //Adding Json body as parameter to the post request
+
+            IRestResponse restResponse = client.Execute(restRequest);
+
+            return restResponse.Content;
```
Comment (Brend-Smits) [inline JobScheduler/Config.cs:38]: Should really return a token model

File: JobScheduler/GraduationWorker.cs [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+    public class GraduationWorker : BackgroundService
+    {
+        private readonly ILogger<GraduationWorker> logger;
+        private List<UserTask> users;
+
+        private readonly ApiRequestHandler requestHandler;
+        public GraduationWorker(ILogger<GraduationWorker> logger)
+        {
+            this.logger = logger;
+            requestHandler = new ApiRequestHandler(new Uri("https://localhost:5001/"));
```
Comment (Brend-Smits) [inline JobScheduler/GraduationWorker.cs:23]: Hardcoded URL

File: Repositories/UserRepository.cs [hunk 13]
```diff
@@ -182,7 +186,16 @@ public bool UserWithRoleExists(Role role)
...
 
+        public async Task<List<User>> GetAllExpectedGraduatingUsers()
+        {
+            DateTime now = DateTime.Now;
+            DateTime max = DateTime.Now.AddMonths(6);
 
-    }
+            return GetDbSet<User>()
+                           .Where(u => u.ExpectedGraduationDate >= now && u.ExpectedGraduationDate <= max)
+                           .ToList();
```
Comment (Brend-Smits) [inline Repositories/UserRepository.cs:196]: return await ..... ToListAsync()

File: Services/2_Services.csproj [hunk 14]
```diff
@@ -5,6 +5,10 @@
     <LangVersion>8</LangVersion>
   </PropertyGroup>
 
+  <ItemGroup>
```
Comment (Brend-Smits) [inline Services/2_Services.csproj:8]: Why is this here?

File: Models/User.cs [hunk 15]
```diff
@@ -31,6 +32,9 @@ public User()
             Services = new List<LinkedService>();
             UserProject = new List<UserProject>();
             FollowedUsers = new List<UserUser>();
+            UserTasks = new List<UserTask>();
```
Comment (Brend-Smits) [inline Models/User.cs:35]: Make sure to also update the User Resource or User ResourceResult if needed.

**Review by Brend-Smits — COMMENTED** [review]
Looks good so far!

**Comment by niraymak** [comment]
Todo:
- Fix migrations
- Send mails if almost graduating
- Secure endpoints. Make them only accessible for JobScheduler
- Make config of jobscheduler return token instead of string?

File: API/Resources/UserTaskResourceResult.cs [hunk 16]
```diff
@@ -0,0 +1,16 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace API.Resources
+{
+    public class UserTaskResourceResult
```
Comment (Maxvanhattum) [inline API/Resources/UserTaskResourceResult.cs:8]: Lacking XML Comments

File: IdentityServer/6_IdentityServer.csproj [hunk 17]
```diff
@@ -14,6 +14,7 @@
           <PrivateAssets>all</PrivateAssets>
           <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
         </PackageReference>
+        <PackageReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Design" Version="3.1.4" />
```
Comment (Maxvanhattum) [inline IdentityServer/6_IdentityServer.csproj:17]: Is this needed in production?

File: IdentityServer/Configuration/IdentityConfig.cs [hunk 18]
```diff
@@ -106,49 +106,74 @@ public static IEnumerable<Client> Clients(Config config)
...
+                                           },
+                       },
+                       new Client
+                       {
+                           ClientId = "dex-jobscheduler",
+                           ClientName = "Digital Excellence Job Scheduler",
+                           AllowedGrantTypes = GrantTypes.ClientCredentials,
+                           ClientSecrets =
+                           {
+                               // new Secret(config.JobScheduler.ClientSecret.Sha256())
```
Comment (Maxvanhattum) [inline IdentityServer/Configuration/IdentityConfig.cs:161]: Commented out unused code.

File: IdentityServer/appsettings.Development.json [hunk 19]
```diff
@@ -38,6 +38,10 @@
         "swagger": {
             "RedirectUrisSwagger": "https://localhost:5001/oauth2-redirect.html",
             "PostLogoutUrisSwagger": "https://localhost:5001"
+        },
+        "JobScheduler": {
+            "ClientId": "",
+            "ClientSecret": ""
```
Comment (Maxvanhattum) [inline IdentityServer/appsettings.Development.json:44]: Does this need to be filled with development data?

File: JobScheduler/ApiRequestHandler.cs [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+            }
+            return userTasks;
+        }
+
+        public void SetGraduationTaskStatusToMailed(UserTask userTask)
+        {
+            RestRequest restRequest = new RestRequest("api/UserTask/SetToMailed") { Method = Method.PUT };
+
+            restRequest.AddParameter("Authorization",
+                                     string.Format("Bearer " + _token),
```
Comment (Maxvanhattum) [inline JobScheduler/ApiRequestHandler.cs:77]: Needs check for if _token is null, just like above.

File: JobScheduler/Config/Config.cs [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+            restClient.Authenticator = new HttpBasicAuthenticator(IdentityServerConfig.ClientId, IdentityServerConfig.ClientSecret);
+
+            IRestResponse identityServerResponse = restClient.Execute(restRequest);
+
+            if(!identityServerResponse.IsSuccessful)
+            {
+
+                Log.Logger.Error("Something went wrong: " + identityServerResponse.ErrorMessage);
+            }
+            Log.Logger.Information(identityServerResponse.Content);
```
Comment (Maxvanhattum) [inline JobScheduler/Config/Config.cs:61]: Is this line necessary for production?

File: JobScheduler/GraduationWorker.cs [hunk 22]
```diff
@@ -0,0 +1,68 @@
+using MessageBrokerPublisher;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using System.Collections.Generic;
+using Models;
+using NotificationSystem.Notifications;
```
Comment (Maxvanhattum) [inline JobScheduler/GraduationWorker.cs:9]: This is a problem with how the current notification system is set up, this is an error since the publisher should not know anything about the consumer (notification system). The fault here at hand is that this is now using the model in the notification system but there is no DTO available in the publisher. I've corrected this mistake in PR: https://github.com/DigitalExcellence/dex-backend/pull/344 .
This should use a DTO available in MessageBrokerPublisher/Models .
Bit of a pain, I know, but necessary if we want to follow the principle of the pub/sub pattern which supplies decoupling.

File: JobScheduler/GraduationWorker.cs [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+        {
+            // Wait till API and Identity are started
+            await Task.Delay(10000);
+
+            while(!stoppingToken.IsCancellationRequested)
+            {
+                logger.LogInformation("Graduation job started: {time}", DateTimeOffset.Now);
+                GraduationJob();
+                logger.LogInformation("Graduation job finished: {time}", DateTimeOffset.Now);
```
Comment (Maxvanhattum) [inline JobScheduler/GraduationWorker.cs:37]: Do we want to log this in production?

File: JobScheduler/GraduationWorker.cs [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+
+                // Time between job. 
+                await Task.Delay(10000, stoppingToken);
```
Comment (Maxvanhattum) [inline JobScheduler/GraduationWorker.cs:40]: Why do we want to wait 10 seconds between every job?

File: JobScheduler/appsettings.json [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+        }
+    },
+    "App": {
+        "API": {
+            "ApiUrl": "https://localhost:5001/"
+        },
+        "IdentityServerConfig": {
+            "IdentityUrl": "https://localhost:5005/",
+            "ClientId": "dex-jobscheduler",
+            "ClientSecret": "dex-jobscheduler"
```
Comment (Maxvanhattum) [inline JobScheduler/appsettings.json:16]: I reckon we want to change this for production to an evironment variable that simultaneously gets used in the Identity Server? This way they always get matched and stay secret.

File: Repositories/UserRepository.cs [hunk 26]
```diff
@@ -185,7 +190,17 @@ public bool UserWithRoleExists(Role role)
                 .SingleOrDefault(r => r.Role.Id == role.Id) != null;
         }
 
+        public async Task<List<User>> GetAllExpectedGraduatingUsers()
+        {
+            DateTime now = DateTime.Now;
+            DateTime max = DateTime.Now.AddMonths(6);
```
Comment (Maxvanhattum) [inline Repositories/UserRepository.cs:196]: I don't think the decision about how many months are added to define expected graduating users should be in the repository. This is a business logic rule and should therefore in my opinion be in the service, passing it to the repository as a parameter.

File: docker-compose.yml [hunk 27]
```diff
@@ -92,6 +96,25 @@ services:
...
+    container_name: jobscheduler
+    restart: on-failure
+    build:
+      context: .
+      dockerfile: JobScheduler/Dockerfile
+    depends_on:
+      - api
+      
+    stdin_open: true
+    tty: true
```
Comment (Maxvanhattum) [inline docker-compose.yml:112]: I don't think these are necessary for service worker template, since it uses a hosted service. Have you tried without?

File: docker-compose.yml [hunk 28]
```diff
@@ -92,6 +96,25 @@ services:
...
+    #networks:
+      #rabbitmq-network:
```
Comment (Maxvanhattum) [inline docker-compose.yml:114]: These need to be not commented out I reckon.

**Review by Maxvanhattum — CHANGES_REQUESTED** [review]

**Comment by Brend-Smits** [comment]
Conflicts, failing checks, also please resolve conversations if they are done. I will review when this is done ^.^

**Comment by niraymak** [comment]
Only 1 Postman test is failing. Couldn't find it last night.
Also have been working on Docker composing the jobscheduler but couldn't finish it yet.
Getting close.

Will look into both issues later.

**Review by MeesvanStraten — APPROVED** [review]

**Comment by matir8** [comment]
Please resolve merge conflicts and unit test warnings https://github.com/DigitalExcellence/dex-backend/actions/runs/501214865

**Review by MeesvanStraten — DISMISSED** [review]

**Review by MeesvanStraten — APPROVED** [review]

**Review by MeesvanStraten — DISMISSED** [review]
